WIP: Improve rendering performance

saml1er opened the pull request:
~~Fixes #905. Texture atlas generator is complete. It combines multiple images together to create a single image.~~ 

See #1061

codenulls committed:
Refactor CElementDeleter for IFP (shared_ptr support) #285

qaisjp committed:
fix typo

codenulls committed:
Add CTextureAtlas and CtextureAtlasInterface

codenulls committed:
Add rectpack2d and DirectXTex to vendor

codenulls committed:
Add CreateTextureAtlas function to CCore

codenulls committed:
Add test code for debugging atlas

codenulls committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	Client/mods/deathmatch/logic/CClientEntity.h
#	Client/mods/deathmatch/logic/CElementDeleter.cpp

codenulls committed:
Use DirectXTex functions to load images

codenulls committed:
include .cpp files in DirectXTex

codenulls committed:
output atlas as png for testing

Samake commented:
Nice to see starts working on this. Thumbs up! What do you think which size atlas can handle later, i mean to hold the whole textures wouldnt be possible. Would it support multiple Atlasses by categories or so?

Einheit-101 commented:
Wow, Sam@ke alive? Nice.

@Topic: I would like to make some performance measurements. I have a lot of time recently.

ghost commented:
> 
> What do you think which size atlas can handle later, i mean to hold the whole textures wouldnt be possible. Would it support multiple Atlasses by categories or so?

You simply need to supply the textures, it will automatically take care of sizes. So basically, nothing will change for scripters except for performance which will make a big difference. Maximum size for a single atlas is 16384x16384. If the textures cannot fit into one atlas then another will be created, so it's up to the program to decide how many atlases are needed, and it will generate them without any issue.

>
>@Topic: I would like to make some performance measurements. I have a lot of time recently.

Sure, once I'm done with the PR. I'll let you know. :P

Citizen01 commented:
You have all the support I can give, just wanted to point out your PR includes 2 commits that are not related to this PR. You should have create a seperate branch on your side to prevent this.

ghost commented:
> 
> You have all the support I can give, just wanted to point out your PR includes 2 commits that are not related to this PR. You should have create a seperate branch on your side to prevent this.

Yes, I realized it later. I'll squash and merge this PR, so these commits will disappear anyway ;p

WADmitry commented:
How about externally prepared textures (using TexturePacker, etc)? For example, then I can use offsets between images and edge bleeding (to avoid alpha/interpolation artifacts on scaled textures).

Also, about maximum size of a single atlas. It depends on hardware. Some of "old" (or just low-end) hardware may not support 16K, 8K and even 4K textures. I think would be good add ability to adjust maximum size (16K/8K/4K).

ghost commented:
> 
> How about externally prepared textures (using TexturePacker, etc)? For example, then I can use offsets between images and edge bleeding (to avoid alpha/interpolation artifacts on scaled textures).

If you're talking about rendering images, then you can always use dxDrawImageSection.  I'm adding the atlas feature only to reduce the state changes (IDirect3DDevice9::SetTexture mainly) for 3D meshes. 

> 
> Also, about maximum size of a single atlas. It depends on hardware. Some of "old" (or just low-end) hardware may not support 16K, 8K and even 4K textures. I think would be good add ability to adjust maximum size (16K/8K/4K).

Yes, we'll need to query the device capabilities and depending on that, we can set the maximum atlas size.

WADmitry commented:
> I'm adding the atlas feature only to reduce the state changes (IDirect3DDevice9::SetTexture mainly) for 3D meshes.

When textures joined into atlas then it may add graphical artifacts on edges (when a textures are independent then the problem usually is not appear) in some cases (for example, when texture is significally zoomed in, interpolation may pick texels of adjacent texture)

ghost commented:
> 
> When textures joined into atlas then it may add graphical artifacts on edges (when a textures are independent then the problem usually is not appear) in some cases (for example, when texture is significally zoomed in, interpolation may pick texels of adjacent texture)

The program will automatically fix the issue for texture mipmaps and remap the 3D mesh UVs properly to avoid edge bleeding. Do you want to do something similar for 2D images that are within an atlas? If yes, then please explain it here.

codenulls committed:
Add DirectXTex .cpp file names to premake5.lua

codenulls committed:
Add CIMGArchive and CIMGArchiveOptimizer

codenulls committed:
Generate custom img archive

codenulls committed:
Merge remote-tracking branch 'origin/master'

codenulls committed:
Add Stdinc to CIMGArhiveOptimizer

codenulls committed:
remove test code from engineReplaceAnimation

codenulls committed:
create window when MTA starts for generating IMG files

codenulls committed:
Make img optimizer window visible on MTA start

codenulls committed:
remove test code from CCore.cpp

codenulls committed:
Add WriteTXD and WriteDFF. Also, fixed RpClumpStreamWrite crashes

RpClumpStreamWrite does not work in GTA SA since they added customized plugins in there, so had to fix a lot of stuff to get it working

codenulls committed:
Add Swap and GetCapacity functions to CBuffer

PlatinMTA commented:
Where this will be launching? On 1.5.7 or on some nightly?

Awesome work btw, MTA got a lot of new cool things these past few months.

ghost commented:
> 
> Where this will be launching? On 1.5.7 or on some nightly?

It will be out on nightlytest.mtasa.com within a month for testing. 1.5.7 will be released soon, so it will come after 1.5.7

ArranTuna commented:
Very nice. If there is a img generator does that mean that only default GTA SA textures are optimizable?

ghost commented:
> 
> 
> Very nice. If there is a img generator does that mean that only default GTA SA textures are optimizable?

No. It will work for all textures and DFF files within the IMG even the custom ones. For DFF, the program will remap the UV coordinates.

ArranTuna commented:
Server modifications aren't in the img though. So wouldn't this mean a total conversion wouldn't have any performance gain?

ghost commented:
> 
> Server modifications aren't in the img though. So wouldn't this mean a total conversion wouldn't have any performance gain?

How about adding a Lua function on client side that can optimize the supplied TXD and DFF files?

ArranTuna commented:
That sounds good. I'm looking forward to testing this BTW.

qaisjp commented on Client/core/CIMGArchive.h:
```diff
@@ -0,0 +1,60 @@
+#ifndef CIMGArchive_H
```
Please use `#pragma once` instead

qaisjp commented on Client/core/CIMGArchive.cpp:
```diff
@@ -0,0 +1,180 @@
+    return blockSize / IMG_BLOCK_SIZE;
+}
+
+#include <iostream>
```
These includes should probably be at top of file

qaisjp commented on Client/core/CIMGArchive.cpp:
```diff
@@ -0,0 +1,180 @@
+{
+    if (archiveFileEntries_.size() <= id || id < 0)
+    {
+        return NULL;
```
Please use `nullptr` wherever possible in all new code.

codenulls committed:
Fix more RpClumpStreamWrite crashes

ghost commented:
`RpClumpStreamWrite` is completely broken in GTA SA. It only works if you disable the custom plugins which rockstar added to the renderware engine, and we cannot do that. I had to rewrite some functions, still 2 more to go for C2dEffect...

codenulls committed:
Add C2dEffectSA

codenulls committed:
Add 2D effect plugin StreamWrite and StreamGetSize callbacks

codenulls committed:
Add GetFileTXDImgArchiveInfo

codenulls committed:
Add CIDELoader for parising .IDE files

codenulls committed:
Parse all IDE sections except cars, hier, and txdp

codenulls committed:
Add SetCurrentDFFBeingGeneratedFileName for debugging

codenulls committed:
Fix DFF plugin related bugs

At this point, all plugins seem to work if you play gta sa with the custom gta3.img. We need 2 callbacks: GetSize and Write for collision plugin in DFF files.

The collision plugin in DFF files is used by vehicles. For now, we're not optimizing vehicle models since we don't have write and getsize callbacks for collision plugin

codenulls committed:
Fix collison structs and classes

codenulls committed:
Parse IDE cars section

codenulls committed:
Rename references to CColData

codenulls committed:
Add collision GetSize and Write callbacks

codenulls committed:
merge vehicle textures with vehicle.txd

and added some debugging stuff

codenulls committed:
Update CBaseModelInfoSAInterface

codenulls committed:
Fix collision plugin bugs and crashes for DFF

ghost commented:
Generating DFF files finally works with support for all plugins. Now, I need to write some more code for creating texture atlases from TXD raster and remapping the geometry vertices to it for DFF files that use it.

We can also add new Lua functions to generate a DFF or COL file for those who wish to export modified COL or DFF. That'll be only useful if we have functions which allow scripters to modify geometry of clump or its collision. Somebody can even create a new ingame editor. @CrosRoad95's collision editing PR can make this a reality.

codenulls committed:
Generate DFF with atlas [Partially working]

Needs a lot of improvement. TODO..

ghost commented:
I was almost about to finish this and found out that the texture must be unwrapped and positioned properly in atlas. Generating it automatically is a very complicated process, which is why human interaction is required to add seams to the mesh where texture might be overlapping and then allow the program to do its job, just like in blender.

Anyways, after spending countless hours, I finally figured out how to automate it without messing up the texture wrapping. We'll need to use [LSCM](https://en.wikipedia.org/wiki/Least_squares_conformal_map), and there is a library named "xatlas" which uses LSCM. However, xatlas was mainly created for baking lightmap, so it doesn't copy the texture to the atlas, I'll need to write some code for that.

Moreover, I removed DirectXTex. It's not needed anymore.

codenulls committed:
Merge branch 'master' into atlases

# Conflicts:
#	Client/core/premake5.lua

codenulls committed:
Remove DirectXTex

codenulls committed:
print .obj file from clump

codenulls committed:
Remove DirectXTex from premake

codenulls committed:
Merge branch 'master' into atlases

codenulls committed:
set toolset to v142 (VS2019)

codenulls committed:
Use xatlas to generate texture atlas

codenulls committed:
Add functions for cloning clump geometries

ghost commented:
The output mesh from xatlas can have more vertices than supplied; hence, we'll need to recreate the geometry for every 3D model. This was a pain because the copying functions for geometry plugins have been optimized out as they are not used, and I had to write my own. It's finally done. 

The new vertices should be copied to the new geometry when it's created, and then we'll have a working build for testers.

codenulls committed:
Fix atlas generator bugs

codenulls committed:
update xatlas

codenulls committed:
generate DFF and TXD with xatlas

Two things must be done now:
1. Fix memory leaks since new textures are created but the previous ones are not destroyed and delete containers allocated using new operator in CIMGArchiveOptimizer.
2. Clumps with a single geometry can be generated, otherwise the program will crash.

codenulls committed:
generate DFF and TXD with xatlas.

Two things must be done now:
1. Fix memory leaks since new textures are created but the previous ones are not destroyed and delete containers allocated using new operator in CIMGArchiveOptimizer.
2. Clumps with a single geometry can be generated, otherwise the program will crash.

PlatinMTA commented:
I really want to test this right now, can't wait.

codenulls committed:
Fix bugs when generating dff with multiple goemetries

Not complete fix. There are still some bugs.

codenulls committed:
Fix DFF incorrect output UV y coordinate

Just needed to flip it.

codenulls committed:
Fix incorrect vertex index xref for clump with multiple geometries

Vehicles are properly generated now. Animated objects look good as well. We still have a problem with material colors for vehicle. All components of vehicles will have the same color, which also includes tyres and windows. Not good at all.

This problem exists because textures for vehicles are handled differently. vehicle.txd in "GTA San Andreas\models\generic" is used as partent txd. We'll probably need to modify some GTA  logic to get it working. Other than that, all other models work fine.

codenulls committed:
Fix TXD memory leak

codenulls committed:
Update xatlas

codenulls committed:
Compress textures to DXT1

codenulls committed:
Fix compiling error

codenulls committed:
Find atlas size for DXT1 compression

codenulls committed:
Move code from CIMGArchiveOptimizer to C3DModelOptimizer

Fixes memory leaks because we are no longer dynamically allocating memory for containers. Memory leak issue still exists for CDXTexture class.

codenulls committed:
Properly calculate atlas DXT1 texture dimensions and size

codenulls committed:
Add GetUsedTexturesCount and GetMostUsedTextureToIgnore

codenulls committed:
Fix GetUsedTexturesCount

codenulls committed:
Create equal # of materials for output geometries

codenulls committed:
Fix typo in GetBestAtlasMaxResolution

codenulls committed:
Ability to ignore repeated textures to atlas to reduce size of TXD

Reduces size by 20% to 60%, depending on which texture we specify to ignore. We'll do this only for models which are bigger in size, like buildings and not for smaller models, like vehicles or peds. The unwrapped texture can be sometimes very big, and we can reduce the atlas size by not adding a texture which is repeated multiple times.

We'll need to prepare a list of one or two textures for every big model (buildings mostly) which should be ignored as the texture might be repeated multiple times resulting in a massive atlas.

codenulls committed:
Use C++ streams to read img files

codenulls committed:
Ignore geometries with no UV data

codenulls committed:
Read multiple files in img at once for better performance

Works sometimes and sometimes it crashes while executing packCharts because of xatlas assertation: XA_DEBUG_ASSERT(r == 0 || r == 1);

codenulls committed:
Get rid of CIMGArchive warnings

codenulls committed:
Get rid of C3DModelOptimizer Warnings

codenulls committed:
Fix output IMG file incorrect data bugs

codenulls committed:
Fix surface memory leak in RwTextureCreateWithFormat

codenulls committed:
Fix all bugs and memory leaks

Everything seems to work fine now except for the line: pRenderWare->WriteTXD(pTXDData, txdArchiveFile.actualFileSize, pAtlasTxdDictionary);

Uncommenting this line will cause a crash. It seems that the crash might be related to actualFileSize being smaller than the write data.

codenulls committed:
Add debugging code

codenulls committed:
Add PackChartsError enum to xatlas (modified vendor)

This will fix xatlas crash which happens for some models.

codenulls committed:
Fix crash when chart doesn't fit into atlas

No more crashes. Tested up to 1925 models from start in gta3.img.

codenulls committed:
Fix typo in CIMGArchive::GetNextImgFiles

filesToRead and entryHeaderIndex variable contained incorrect values if the fuction is called more than once

codenulls committed:
Add CDFFModelOptimizationInfo

codenulls committed:
code cleanup

codenulls committed:
Fix ReadTextureDictionary return typo

codenulls committed:
Fix memory leak from previous commits

codenulls committed:
output txd and dff optimization info to .GMOI file

GMOI = GTA Model Optimization Info

codenulls committed:
Add SetupAtlasDataForOptimizationInfo function

codenulls committed:
Add functions for finding model optimized data

codenulls committed:
Fix crash from previous commit

ghost commented:
Let's just say that I'm very close to finish this. I tried generating atlases for all models and they all worked fine except for 3-4 due to some problems with xatlas. No big deal, I'll contact the author, and hopefully he'll fix it. He's actively developing xatlas. You can check my to-do list to see what's left.

I came across low LOD models, and it seems that we cannot use the LOD models from GTA SA if we are going to use atlases. Technically, we can use them, but when the model is not fully visible, the low LOD model will be displayed, and it will use multiple draw calls since every low LOD model uses a different TXD file which has textures with extremely low resolution. We cannot generate atlases for these low LOD models because then the textures will look like huge pixelated color boxes.  

What are we going to do then? I'm going to use another library for generating low LOD models from the new atlas DFF and TXD files. We need LOD support for our players who are running GTA on low-end computers. I'll do this when I finish working on atlases. 

To-do list:

- [x] Avoid adding repeated textures to atlas to reduce atlas size (needs automation)
- [ ] Patch some GTA code to fix messed up color bug with vehicle models. FYI, a vehicle model on screen will take 15 to 24 draw calls per frame. We are reducing it to 1.
- [ ] Rewrite some renderware rendering functions to support atlases, since without doing that, we'll see absolutely no change. VERY IMPORTANT!
- [ ] Use a 3rd party library to generate low lod models instead of using the GTA ones. Low priority.

ArranTuna commented:
Maybe low LOD's won't even be necessary, when those renderware functions are rewritten and you've seen the performance improvement, maybe it will be so high that the difference of having low LOD's will be insignificant and instead the full model could just be displayed?

Einheit-101 commented:
GTA LOD system is just annoying anyways, and even the default models are so low poly, they are insignificant.

codenulls committed:
Update xatlas (all crashes fixed)

The vendor is no longer modified because the author fixed the crashes.

codenulls committed:
Ability to read and write to GMOI files properly

GMOI = GTA Model Optimization Info

GMOI files contain data about texture sizes within atlas for the model. It defines how much size each texture will take within the atlas.

ghost commented:
@ArranTuna It's not just about performance. It also saves you lots of memory if a model is far away, GTA unloads it and shows the low LOD. Like I said, LOD models have their own TXD files. These TXD files will take up only a few KBs and save memory if you are not using custom models. 

I do have plans to add LOD support for custom models after this PR. 

@Einheit-101 Maybe you're right, but one can argue that if you are flying a hydra and you position your camera in a way where you see thousands of objects on your screen, there will be some performance issue on weak computers.

ArranTuna commented:
> I do have plans to add LOD support for custom models after this PR.

If you mean objects that servers have applied mods to, there is already some functions for that like setLowLODElement which I have used on modified land masses (but I just use the exact same texture instead of wasting the time and another model ID slot for a low quality texture).

codenulls committed:
Avoid adding repeated textures to atlas

Added some basic code for ignoring repeated textures. It will work, however, improvement is needed.

codenulls committed:
Add ide-ipl-lib library (modified)

For reading and writing IDE files.

codenulls committed:
Use ide-ipl-lib library to read and write IDE files

codenulls committed:
Update xatlas

codenulls committed:
Split new IMG into multiple img files

NOT TESTED. If we keep everything in RAM while generating atlases, we might run out of memory or everything might slow down and take too much time to complete. Later, we'll need to merge all IMG files into one when we are done writing all new DFF and TXD files.

codenulls committed:
output new img files to the output folder

codenulls committed:
Move atlas size reducer code to CAtlasSizeReducer

This code should be refactored..

codenulls committed:
fix a warning

codenulls committed:
Reduce atlas size in CAtlasSizeReducer

codenulls committed:
Fix corrupt atlas TXD names in output IMG files

codenulls committed:
Merge branch 'master' into atlases

# Conflicts:
#	Client/core/CCore.h

codenulls committed:
Fix RwD3D9RasterCreate leak

codenulls committed:
Fix compiling issue

codenulls committed:
Fix crash with incorrect atomics in list

codenulls committed:
Merge all output img files into a final one

codenulls committed:
Set maximum charts pixels for atlas

codenulls committed:
Revert to old xatlas version to fix crashes

Reverted to this commit: https://github.com/jpcy/xatlas/commit/999bbf3eebb9359dc9f4631370e80f7084134491

codenulls committed:
Fix line endings for CRenderWareSA.cpp from 7478e1a4b3f7bacfec07ef3a9c4ed93071fcdbd4.

codenulls committed:
Patch SA IDE paths to load atlas DFF and TXD files from IMG

There's a crash when loading some specific model. We need to find out the model causing it.

codenulls committed:
Add debugging code to find problematic model causing crash

ghost commented:
I was able to generate a custom img file with atlas models. It was working, however, I could not load more than 4500 TXD files since that's the limit. The game kept crashing when I tried to exceed the limit, so I'll have to fix this by merging TXD files. I wanted to lift the model ID limit, but I got busy with this PR.

Just two more important things to actually benefit from the atlases: 

- Rewrite some renderware rendering functions to support atlases
- Patch some GTA code to fix messed up color bug with vehicle models

codenulls committed:
Merge TXD files together to avoid exceeding limit and crash

Tested. All models work. Alpha and color bug needs fixing.

ghost commented:
All models seem to work, but merging TXD files is not a good idea. It results in TXD files being large. I limited the atlas TXD size to 2mb, and I still got 1200 atlas TXDs. I cannot have more than 1500 custom TXD files since GTA internal TXD names will take up the remaining 3500. When you are flying hydra, the streaming memory usage will be very high and objects will start to disappear.  

It happens because loading a single DFF model and its TXD will result in unwanted textures being loaded along with the TXD, and it will eat lots of memory. That's not really what we are looking for. There are 8000 DFF models which are worth optimizing, so we need to uplift the limit to **12000 TXD files** to make the streamer run smooth and use less memory. I'm going to use some code from fastman92 limit adjuster to make sure every DFF with atlas can have its own TXD.

codenulls committed:
Add Fastman92 limit adjuster code

ghost commented:
I've decided to stop working on this, and go for DX11 instead. It's going to take a lot of work to implement the limit adjuster and fix the bugs/crashes that might appear. Moreover, the atlases textures will eat some good amount of RAM, and the only way to fix that is to reduce the texture sizes, but that's not what we want. Like I said, too much work to make it happen, why not spend that time on implementing upgrade to DX11? 🤔🤔🤔🤔🤔🤔

The atlases code is completely useless now. However, there is another feature in this PR, and that is the ability to output COL, DFF, and TXD files. Basically, we can add some new functions to allow scripters to create their own ingame editors for modifying DFF clumps, textures, and COL models, and then output them to a file.

Let's talk in #1061
